doc/coding-guidelines.md

```diff
@@ -0,0 +1,447 @@
+> Why? Otherwise, there is no guarantee that new colors don't look alien for a random VS Code color theme.
+<a name="theming-theia-colors"></a>
+* [4.](#theming-theia-colors) Apply different color values only in concrete Theia themes, see [Light (Theia)](https://github.com/eclipse-theia/theia/blob/master/packages/monaco/data/monaco-themes/vscode/light_theia.json), [Dark (Theia)](https://github.com/eclipse-theia/theia/blob/master/packages/monaco/data/monaco-themes/vscode/dark_theia.json) and [High Contrast (Theia)](https://github.com/eclipse-theia/theia/blob/master/packages/monaco/data/monaco-themes/vscode/hc_theia.json) themes.
+<a name="theming-variable-naming"></a>
+* [5.](#theming-variable-naming) Names of variable follow the `object.property` pattern.
+```ts
+// bad
+'button.secondary.foreground'
+'button.secondary.disabled.foreground'
+
+// good
+'secondaryButton.foregrond'
+'secondaryButton.disabledForegrond'
+```
+
+## React
+<a name="no-bind-fn-in-event-handlers"></a>
+* [1.](#no-bind-fn-in-event-handlers) Do not bind functions in event handlers.
+  - Extract a react component if you want to pass state to an event handler function.
+
+> Why? Because it creates a new instance of event handler function on each render and breaks React element caching leading to rerendering and bad performance.
+
+```ts
+// bad
+class MyWidget extends ReactWidget {
+  render(): React.ReactNode {
+    return <div onClick={this.onClickDiv.bind(this)} />;
+  }
+
+  protected onClickDiv(): void {
+    // do stuff
+  }
+}
+
+// bad
+class MyWidget extends ReactWidget {
+  render(): React.ReactNode {
+    return <div onClick={() => this.onClickDiv()} />;
+  }
+
+  protected onClickDiv(): void {
+    // do stuff
+  }
+}
+
+// very bad
+class MyWidget extends ReactWidget {
+  render(): React.ReactNode {
+    return <div onClick={this.onClickDiv} />;
+  }
+
+  protected onClickDiv(): void {
+    // do stuff, no `this` access
+  }
+}
+
+// good
+class MyWidget extends ReactWidget {
+  render(): React.ReactNode {
+    return <div onClick={this.onClickDiv} />
+  }
+
+  protected onClickDiv = () => {
+    // do stuff, can access `this`
+  }
+}
+```
+
+## URI/Path
+<a name="uri-over-path"></a>
+* [1.](#uri-over-path) Pass URIs between frontend and backend, never paths. URIs should be sent as strings in JSON-RPC services, e.g. `RemoteFileSystemServer` accepts strings, not URIs.
+> Why? Frontend and backend can have different operating systems leading to incompatibilities between paths. URIs are normalized in order to be os-agnostic.
+<a name="frontend-fs-path"></a>
+* [2.](#frontend-fs-path) Use `FileService.fsPath` to get a path on the frontend from a URI.
+<a name="backend-fs-path"></a>
+* [3.](#backend-fs-path) Use `FileUri.fsPath` to get a path on the backend from a URI. Never use it on the frontend.
```

Comment:
@msujew, these items also fall into your expertise - what's your take on this advice?